queue/QueueConsumer.js

```diff
@@ -24,13 +52,57 @@ module.exports = class QueueConsumer extends BaseQueue {
     if (!this._channel) {
       throw new Error('Channel not initialized')
     }
+
+    const messageContent = message.content
+
+    const maxMessageRetries = this._maxMessageRetries
+    const deadLetterQueueName = this._deadLetterQueueName
+
+    const { headers } = message.properties
+    const deadLetterHeader = headers[DEAD_LETTER_HEADER]
+
+    let rejectedRecord
+    if (deadLetterHeader) {
+      // find 'rejected' record
+      // TODO: handle other reasons in the future
+      // ex. dead letter messages from TTL expires (not implemented yet)
+      for (const record of deadLetterHeader) {
+        if (record.reason === 'rejected') {
+          rejectedRecord = record
+        }
+      }
+    }
+
+    if (rejectedRecord) {
+      // get the number of times this message has been attempted
+      const attemptCount = rejectedRecord.count
+
+      if (attemptCount < maxMessageRetries) {
+        // send to dead letter exchange again
+        return this._channel.nack(message, false, false)
+      }
+
+      // otherwise...
+      // publish to dead letter queue to be handled later
+      await this._channel.sendToQueue(deadLetterQueueName, messageContent)
+
+      // acknowledge message so it doesn't show up on
+      // queue again
+      return this._channel.ack(message)
+    }
+
     // reject message, don't reject all messages,
-    // requeue the message
-    return this._channel.nack(message, false, true)
+    // and don't requeue it in the main queue
+    // (this will move the message over to the dead letter exchange
+    // and will cause the message to be handled by instances of this
+    // consumer again)
+    return this._channel.nack(message, false, false)
```

Comment:
I think we should make the `false` and `false` constants at the top of the file or at least put a comment mentioning what they mean. I probably won't remember what those actually mean. It would also keep the above reference to `nack` consistent as well.